lib/src/front_end/comment_writer.dart

```diff
@@ -0,0 +1,277 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/source/line_info.dart';
+
+import '../piece/sequence.dart';
+import '../source_comment.dart';
+import 'piece_writer.dart';
+
+/// Functionality used by [AstNodeVisitor] to build text and pieces from the
+/// comment tokens between meaningful tokens used by AST nodes.
+///
+/// Also handles preserving discretionary blank lines in places where they are
+/// allowed.
```

Comment:
This is surprising - can you expand this a bit to indicate why comments and blank lines are similar and worth implementing together?

Reply:
Good point, done. (The old formatter worked this way too, so I didn't think to write it down.)